### DISCUSSION TITLE
Vercel.json's "includeFiles" with `pages/api` serverless functions

### DISCUSSION THREAD

**Discussion by zackkrida** [vercel/next.js#14807] [Help] [answered]
Lets say i had two files in the root of my next.js project: `random.txt` and `somedir/random2.txt`. How could I access those inside of a `src/pages/api/function.js` file? Whatever I do I can't seem to get the paths right or even verify that the files are present. I thought something like the below would help:

```json5
// vercel.json
{
  "functions": {
    "src/pages/api/function.js": {
      "includeFiles": "somedir/random2.txt"
    }
  }
}
```

But no luck.

**Comment by Timer** [comment]
Can you share your project code? There's zero configuration necessary for this.

**Reply by fozziethebeat** [reply]
How did this get resolved? I'm trying to do this with a readonly sqlite database but I can't access it when running in prod.

**Reply by jsbimra** [reply]
Same situations here, I wrote to vercel support team 6 days ago, hoping to see some solution from them.

**Reply by QB3L** [reply]
Where is the answer exactly?

**Reply by OR13** [reply]
I used this: https://github.com/vercel/next.js/issues/8251#issuecomment-927114789

**Reply by QB3L** [reply]
Gotcha, also need to specify that zero config to make this work is actually on next 12

**Reply by Solomon04** [reply]
How the hell does this answer the question?

**Reply by QB3L** [reply]
@Solomon04 take a look at https://github.com/vercel/next.js/issues/8251

I tried to summarize the discussion as well on one of my comments:
https://github.com/vercel/next.js/discussions/32236#discussioncomment-1955080

**Reply by callumlocke** [reply]
2 years later, this is still here, still a problem, still marked as an Answer when it contains no answer.

**Reply by AlonMiz** [reply]
frustrated with this for almost a day now. tried EVERYTHING. still no luck, I'm using luck because it's exactly that. there's no good indication or feedback for this. mostly trial and error
(using next.js version 12.3.1)

## vercel.json function includeFiles - FAIL

from vercel.json, every combination possible (was getting a lot of `The pattern "render/**/*" defined in `functions` doesn't match any Serverless Functions.` but eventually, it did accept the config) with glob. tried even to glob everything.
```
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "functions": {
    "pages/api/render/[accountId].ts": {
      "includeFiles": "**/*"
    }
  }
}
```
UPDATE: apparently `includeFiles` prop in `functions` doesn't work with next.js. i got this error after accidentally deploying via the vercel CLI. 
I haven't seen this error either in the UI or with `vercel build` or `vercel dev`
> ❗️  Ignoring function property `includeFiles`. When using Next.js, only `memory` and `maxDuration` can be used.
> Learn More: https://vercel.link/functions-property-next
> 

after 20 commits and push, I used the vercel cli which resulted in the same unexpected result. nothing was included in the function.

## Page config with unstable_includeFiles - FAIL
tried this config, but it didn't work as well.

```
export const config: PageConfig = {
  unstable_includeFiles: ['src'],
};
```

## Auto Detection - FAIL
tried to include some extra files using the `zero config` / `tracing technique` / `@vercel/nft` logic (don't get me started pls)
but it was able to catch only one file of the tree i needed.
`const pathToComponent = path.join(process.cwd(), './src/components/Page.tsx');`
stating here another very odd behavior
same folder:
```
  const a = path.join(process.cwd(), './src/components/a.tsx');
  const b = path.join(process.cwd(), './src/components/b.tsx');
  const c = path.join(process.cwd(), './src/components/c.tsx');
  const d = path.join(process.cwd(), './src/components/d.css');
```
only `a.tsx` and `d.css` are copied. and my assumption is that the files that are being copied are files that are not part of the original webpack build. so I'm left only with files that are not bundled. 😞 

I'm pretty frustrated on my first day using vercel.

## Netlify - SUCCESS
stating here that netlify does work with a simple config file. simply adding this to netlify.toml

```
[functions]
  included_files = ["src/**"]
```

**Reply by AlonMiz** [reply]
i was able to reproduce it in this repo
https://github.com/AlonMiz/next-js-reproduce
this is the app
https://next-js-reproduce.vercel.app/

**Reply by thevinci** [reply]
So is the only solution go use Netlify? WTF.

**Reply by wiiiimm** [reply]
@thevinci @AlonMiz while `excludeFiles` and `includeFiles` (in `vercel.json`) doesn't work for Next.js, there are `experimental.outputFileTracingExcludes` and `experimental.outputFileTracingIncludes`.

You can try `experimental.outputFileTracingIncludes` by having the following in `next.config.js`:

```
module.exports = {
  experimental: {
    outputFileTracingExcludes: {
      '/api/hello': ['./un-necessary-folder/**/*'],
    },
    outputFileTracingIncludes: {
      '/api/another': ['./necessary-folder/**/*'],
    },
  },
}
```
Source: https://nextjs.org/docs/pages/api-reference/next-config-js/output#caveats

**Comment by dipunj** [comment]
Anyone else stumble upon this? Here is something you should try out, especially if you are reading a local file for an api call.
It could be because you have referenced it directly as:

`fs.readFileSync("PostList.json")`


Vercel deployments are serverless and to ensure that at run time the path is correct, please use the following:

`fs.readFileSync(process.cwd(), "PostList.json")`

**assuming your file `PostList.json` is located at the project root directory.

**Comment by gendronb** [comment]
@Timer here's a sample NextJS project I can share with you (https://github.com/gendronb/next-bullmq). 

The project works perfectly locally.

My goal is to include .LUA files from node_modules (required by BullMQ, see https://github.com/taskforcesh/bullmq/issues/469#issuecomment-832647151), but in the sample code provide, I simply try to include some files located at the project root.

When I deploy to Vercel and I inspect the deployment output, these files are not present.

I did try many variations of vercel.json, with no luck.

**Comment by hypervillain** [comment]
I have the same issue. Things work locally but trigger an internal server error on Vercel.
Here is a simple repro repository: https://github.com/hypervillain/import-files-repro

Any help would be greatly appreciated 😊

**Comment by chris-hinds** [comment]
I am having the exact same issue, I have a private key file in the root directory that an API function requires. The function works fine locally. However inspecting the vercel built output the private key file is nowhere to be see.

I appreciate loading a private key file in this way is insecure...... I am switching to including the contents via an env variable. I just wanted to provide another example.

**Comment by pedroresende** [comment]
Any news about this? Also having the same problem

**Reply by ijjk** [reply]
Hi, can you provide additional info/a minimal reproduction for where you are seeing a need for this?

**Comment by OR13** [comment]
surprisingly difficult to do basic file reading... 

Came here after seeing this deceptively simple (not working) answer: 


```
// pages/api/image.js

import fs from 'fs'
import path from 'path'

const filePath = path.resolve('.', 'images_folder/next.jpg')
const imageBuffer = fs.readFileSync(filePath)

export default function(req, res) {
  res.setHeader('Content-Type', 'image/jpg')
  res.send(imageBuffer)
}
```

https://stackoverflow.com/questions/63066985/send-file-as-response-using-nextjs-api

Make sure you install the correct version of next described here: https://github.com/vercel/next.js/issues/8251#issuecomment-927114789

**Comment by diegoulloao** [comment]
Why not to put all included files into just one folder? `someDir`

Try:

`vercel.json`
```json
{
  "functions": {
     "pages/api/function.js":  {
       "includeFiles": "someDir/*.txt"
     }
  }
}
```

Where pattern `function.js` is a route handler inside `pages/api`. 
For **vercel serverless functions** just use `api/function.js`

`pages/api/function.js`
```js
// ... rest of handler

const txtFile1 = fs.readFileSync(
  path.resolve(process.cwd(), 'someDir/random.txt'),
  { encoding: 'utf-8' }
)

const txtFile2 = fs.readFileSync(
  path.resolve(process.cwd(), 'someDir/random2.txt'),
  { encoding: 'utf-8' }
)
```

If you wanna get access to those files in other routes handlers too, you can try this (for all):

`vercel.json`
```json
{
  "functions": {
     "page/api/*.js":  {
       "includeFiles": "someDir/*.txt"
     }
  }
}
```

Just use [glob patterns](https://github.com/isaacs/node-glob#glob-primer).
Please check: https://vercel.com/docs/project-configuration#project-configuration/functions

Also these testers may help:
- https://www.digitalocean.com/community/tools/glob
- https://globster.xyz/

It's super important to use `process.cwd()` instead of `__dirname` or `.` to get the paths to the files.
`cwd()` will return the directory where the project is being executed.

If you have multiple functions rules then **order matters**.

**Reply by apecollector** [reply]
This doesn't work because of:

```
❗️  Ignoring function property `includeFiles`. When using Next.js, only `memory` and `maxDuration` can be used.
```

**Reply by wiiiimm** [reply]
while `excludeFiles` and `includeFiles` (in `vercel.json`) doesn't work for Next.js, there are `experimental.outputFileTracingExcludes` and `experimental.outputFileTracingIncludes`.

You can try `experimental.outputFileTracingIncludes` by having the following in `next.config.js`:

```
module.exports = {
  experimental: {
    outputFileTracingExcludes: {
      '/api/hello': ['./un-necessary-folder/**/*'],
    },
    outputFileTracingIncludes: {
      '/api/another': ['./necessary-folder/**/*'],
    },
  },
}
```
Source: https://nextjs.org/docs/pages/api-reference/next-config-js/output#caveats

**Comment by Oisangui** [comment]
What worked for me was to copy what you need to the assets folder with a custom `build` script. Those files will be available to your API route. 

For my usecase I needed to compile and type-check typescript code dynamically in the backend but the files I needed were not available in run-time so I added them to the assets folder. The typescript compiler goes looking for some lib files in the node_modules folder, so to circumvent that I use a virtual filesystem and voila.

**Reply by anzksdk** [reply]
hi, can you provide the build script?
I dont know the source and destination of the cp command
thank you very much

**Reply by nmitic** [reply]
This the route I am taking as well after trying all the workaround here that did not worked for me, however, even after being able to copy files that I need, I am still getting not found error from vercel...

**Reply by bxclib2** [reply]
I feel crazy why they do not help really solve the problem when everyone is stucking into this. They keep releasing nextjs, making more and more bugs in the recent versions.

**Comment by FrancescoSaverioZuppichini** [comment]
same here

**Comment by leerob** [comment, marked as answer]
I updated our guide with more documentation and examples here for different frameworks. Hope this helps!

https://vercel.com/guides/how-can-i-use-files-in-serverless-functions

**Reply by rtritto** [reply]
@leerob I'm using Vike (Vite based) and I have some files in `/server/api` directory that are correctly bundled (during the build step) into `/dist/server/api` directory with `mjs` extension.
In `/dist/server/index.mjs` file, I use `await import('./api/<FUNCTION>.mjs')`.

In `vercel.json` I added:
```diff
{
  "outputDirectory": "dist/client",
+  "functions": {
+    "api/ssr.js": {
+      "includeFiles": "dist/server/api/*.mjs"
+    }
+  },
  "rewrites": [
    {
      "source": "/((?!assets/).*)",
      "destination": "/api/ssr.js"
    }
  ]
}
```

But I get:
```console
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/dist/server/api/<FUNCTION>.mjs' imported from /var/task/dist/server/index.mjs
    at finalizeResolution (node:internal/modules/esm/resolve:263:11)
    at moduleResolve (node:internal/modules/esm/resolve:919:10)
    at moduleResolveWithNodePath (node:internal/modules/esm/resolve:1043:14)
    at defaultResolve (node:internal/modules/esm/resolve:1086:79)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:650:12)
    at #cachedDefaultResolve (node:internal/modules/esm/loader:599:25)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:582:38)
    at ModuleLoader.getModuleJobForImport (node:internal/modules/esm/loader:241:38)
    at onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:542:36)
    at TracingChannel.tracePromise (node:diagnostics_channel:337:14) {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///var/task/dist/server/api/<FUNCTION>.mjs'
}
```

How can I fix this issue?

**Reply by johannes-lindgren** [reply]
@rtritto, I was also struggling doing my own build (with ESBuild) until I found this article:

https://vercel.com/blog/build-your-own-web-framework#-serverless-functions

And via that, this example repo:

https://github.com/vercel/examples/tree/main/build-output-api/serverless-functions/.vercel/output/functions/index.func

What I needed to add to my build step:

- Delete `vercel.json`: apparently, it's a build configuration, but when I build myself, I don't need it. 
- Output to `.vercel/output/functions/index.func` (instead of `dist`)
- Create a `.vercel/output/config.json` file
- Create a `.vercel/output/functions/index.func/.vc-config.json` file with the following content


```json
{
  "runtime": "nodejs22.x",
  "handler": "index.mjs",
  "launcherType": "Nodejs"
}
```

In my case, I was using ES modules so I had to use the `.mjs` file extension 

But this was not summarized anywhere, so I struggled even after I found this GitHub issue.
